pkg/controller/cainjection/apiservice_controller.go

```diff
@@ -0,0 +1,334 @@
+	APIServiceList := &apiregistrationv1.APIServiceList{}
+	err := c.List(context.TODO(), APIServiceList, &client.ListOptions{})
+	if err != nil {
+		log.Error(err, "unable to list apiservices")
+		return []apiregistrationv1.APIService{}, err
+	}
+	result := []apiregistrationv1.APIService{}
+	for _, APIService := range APIServiceList.Items {
+		if ann, ok := APIService.GetAnnotations()[certAnnotationServiceCA]; ok && ann == "true" {
+			result = append(result, APIService)
+		}
+	}
+	return result, nil
+}
+
+type enqueueRequestForReferecingAPIServices struct {
+	client.Client
+	InjectionType string
+}
+
+// trigger a router reconcile event for those routes that reference this secret
+func (e *enqueueRequestForReferecingAPIServices) Create(evt event.CreateEvent, q workqueue.RateLimitingInterface) {
+	APIServices := []apiregistrationv1.APIService{}
+	if e.InjectionType == secretInjection {
+		APIServices, _ = matchSecretWithAPIService(e.Client, types.NamespacedName{
+			Name:      evt.Meta.GetName(),
+			Namespace: evt.Meta.GetNamespace(),
+		})
+	}
+	if e.InjectionType == systemCAInjection {
+		APIServices, _ = matchSystemCAWithAPIService(e.Client)
+	}
+	for _, CRD := range APIServices {
+		q.Add(reconcile.Request{NamespacedName: types.NamespacedName{
+			Name: CRD.GetName(),
+		}})
+	}
+}
+
+// Update implements EventHandler
+// trigger a router reconcile event for those routes that reference this secret
+func (e *enqueueRequestForReferecingAPIServices) Update(evt event.UpdateEvent, q workqueue.RateLimitingInterface) {
+	APIServices := []apiregistrationv1.APIService{}
+	if e.InjectionType == secretInjection {
+		APIServices, _ = matchSecretWithAPIService(e.Client, types.NamespacedName{
+			Name:      evt.MetaNew.GetName(),
+			Namespace: evt.MetaNew.GetNamespace(),
+		})
+	}
+	if e.InjectionType == systemCAInjection {
+		APIServices, _ = matchSystemCAWithAPIService(e.Client)
+	}
+	for _, CRD := range APIServices {
+		q.Add(reconcile.Request{NamespacedName: types.NamespacedName{
+			Name: CRD.GetName(),
+		}})
+	}
+}
+
+// Delete implements EventHandler
+func (e *enqueueRequestForReferecingAPIServices) Delete(evt event.DeleteEvent, q workqueue.RateLimitingInterface) {
+	return
+}
+
+// Generic implements EventHandler
+func (e *enqueueRequestForReferecingAPIServices) Generic(evt event.GenericEvent, q workqueue.RateLimitingInterface) {
+	APIServices := []apiregistrationv1.APIService{}
+	if e.InjectionType == secretInjection {
+		APIServices, _ = matchSecretWithAPIService(e.Client, types.NamespacedName{
+			Name:      evt.Meta.GetName(),
+			Namespace: evt.Meta.GetNamespace(),
+		})
+	}
+	if e.InjectionType == systemCAInjection {
+		APIServices, _ = matchSystemCAWithAPIService(e.Client)
+	}
+	for _, CRD := range APIServices {
+		q.Add(reconcile.Request{NamespacedName: types.NamespacedName{
+			Name: CRD.GetName(),
+		}})
+	}
+}
+
+func (r *ReconcileAPIService) getSecretCA(secretName string, secretNamespace string) ([]byte, error) {
```

Comment:
this seems to be duplicated code as we keep adding new controllers, we should probably look to create a refactor task for things like this

Reply:
I had the same thought. We should add a task to refactor all the repeated code, but I don't think that should stop this merge.